### PR TITLE
Boosting Algorithms in Python

### LINKED ISSUE #2503: Introduction to Boosting Algorithms
Boosting algorithms have taken over almost all Kaggle notebooks. They have found their way into almost every award-winning model and have been the most sought-after techniques in Machine Learning. In this article, the reader is introduced to boosting algorithms, how they work and how to use them.

Key Takeaways:
1. Boosting algorithms
2. How they work
3. Examples of boosting algorithms
4. Using boosting algorithms to product results

References:
N/A

### PR DESCRIPTION
**Pull request by bloodprashure**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

This fixes #2503

### CODE REVIEW

**Comment by bloodprashure** [comment]
This pull fixes #2503

**Comment by bloodprashure** [comment]
This fixes #2503

File: content/articles/boosting-algorithms-part-1/index.md [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+4. [AdaBoost](#adaboost)
+5. [AdaBoost Implementation in Python](#adaboost-implementation-in-python)
+6. [XGBoost](#xgboost)
+7. [XGBoost Implementation in Python](xgboost-implementation-in-python)
+8. [Additional Resources](#additional-resources)
+
+### Introduction
+The first-ever boosting algorithm called AdaBoost (Adaptive Boosting) was introduced to the world by Freund and Schapire. [Statisticians](https://web.stanford.edu/~hastie/Papers/buehlmann.pdf) have come forward, laying proofs and theorems as to how this gradient descent algorithm in function space, based on statistical estimations and numerical optimizations has proven to be very competitive in terms of prediction accuracy.
+
+AdaBoost was described as a "stagewise, additive modeling", where "additive" didn't mean a model fit added by covariates, but meant a linear combination of estimators. An ensemble technique almost always confused with bagging, boosting can be considered as extensively regularizing a model through model iterations.
```
Comment (srishilesh) [inline content/articles/boosting-algorithms-part-1/index.md:35]: This intro looks a little technical. Beginners would find it difficult to understand.

Instead of directly jumping into Adaboost, you could probably talk a little about Ensemble methods, and boosting.

File: content/articles/boosting-algorithms-part-1/index.md [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+import numpy as np
+import pandas as pd
+import sklearn
+import xgboost
+```
+
+### Bagging vs Boosting
+As mentioned, boosting is confused with [bagging](https://en.wikipedia.org/wiki/Bootstrap_aggregating). Those are two different terms, although both are ensemble methods. Let us try to understand what ensembling is, and move on to breaking down the differences between bagging and boosting.
+
+[Ensemble methods](https://www.section.io/engineering-education/ensemble-learning/) are techniques that use multiple models and combine them into one for enhanced results. The term "models" could refer to any model - regression, support vector machines, and kNNs, and the model whose performance has to be improved is called the base model. Although the technique "boosting" uses decision trees to improve the model's accuracy, it can be applied to any base model.
```
Comment (srishilesh) [inline content/articles/boosting-algorithms-part-1/index.md:66]: You could move this to the intro

File: content/articles/boosting-algorithms-part-1/index.md [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+```py
+import numpy as np
+import pandas as pd
+from sklearn.ensemble import AdaBoostClassifier
+from sklearn.model_selection import train_test_split
+from sklearn.preprocessing import LabelEncoder
+from sklearn.preprocessing import StandardScaler
+```
+
+Now, let's import the dataset using the `read_csv()` method in Pandas and analyze the number of distinct categories in each feature. If a feature has only one unique value, we can drop it, as it has no significance while building the model.
```
Comment (srishilesh) [inline content/articles/boosting-algorithms-part-1/index.md:96]: You could insert an image containing the dataframe's head() (Top 5 rows) of the dataset

File: content/articles/boosting-algorithms-part-1/index.md [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+As you can see `Number of unique values in veil-type is 1`, the feature `veil-type` has only one distinct value in it and hence, can be dropped.
+
+```py
+df = df.drop("veil-type",axis=1)
+```
+
+Since machine learning models prefer numerical data, let us convert the dataset to numbers by encoding it.
+
+```py
+label_encoder = LabelEncoder()
```
Comment (srishilesh) [inline content/articles/boosting-algorithms-part-1/index.md:140]: Give 1 - 2 lines of explanation of how `LabelEncoder()` works. And, add any external link that the reader might find it useful to understand more about label encoding.

File: content/articles/boosting-algorithms-part-1/index.md [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+from sklearn.preprocessing import LabelEncoder
+from sklearn.preprocessing import StandardScaler
+```
+
+Now, let's import the dataset using the `read_csv()` method in Pandas and analyze the number of distinct categories in each feature. If a feature has only one unique value, we can drop it, as it has no significance while building the model.
+
+```py
+df = pd.read_csv("mushrooms.csv")
+for col in df.columns:
+    print('Number of unique values in',col,'is',len(df[col].unique()))
```
Comment (srishilesh) [inline content/articles/boosting-algorithms-part-1/index.md:101]:
```suggestion
    print('Number of unique values in' ,col, 'is', len(df[col].unique()))
```

File: content/articles/boosting-algorithms-part-1/index.md [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+Number of unique values in ring-type is 5
+Number of unique values in spore-print-color is 9
+Number of unique values in population is 6
+Number of unique values in habitat is 7
+```
+
+As you can see `Number of unique values in veil-type is 1`, the feature `veil-type` has only one distinct value in it and hence, can be dropped.
+
+```py
+df = df.drop("veil-type",axis=1)
```
Comment (srishilesh) [inline content/articles/boosting-algorithms-part-1/index.md:134]:
```suggestion
df = df.drop("veil-type", axis=1)
```

File: content/articles/boosting-algorithms-part-1/index.md [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+```py
+label_encoder = LabelEncoder()
+for column in df.columns:
+    df[column] = label_encoder.fit_transform(df[column])
+```
+
+Splitting the dataset into a target matrix `Y` and a feature matrix `X`,
+
+```py
+X = df.drop('class',axis=1)
```
Comment (srishilesh) [inline content/articles/boosting-algorithms-part-1/index.md:148]: Generally, it is not the right approach to drop a column from the main data frame. If you drop, the original data in the data frame is lost.

Instead, slice the columns that you require for `X` and `Y`.

File: content/articles/boosting-algorithms-part-1/index.md [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+Y = df['class']
+```
+
+The dataset must be split into two - training data and testing data. Let us go ahead and split the data, 70% of it for training and 30% for testing and standardize the values.
+
+```py
+X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state=100)
+scaler = StandardScaler()
```
Comment (srishilesh) [inline content/articles/boosting-algorithms-part-1/index.md:156]: Explain why use `StandardScaler()`

File: content/articles/boosting-algorithms-part-1/index.md [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+
+These decision stump algorithms are used to identify weak learners. Starting with one decision tree, the misclassified examples are penalized by increasing their weight (the weight is boosted), and another decision tree is built from the new and modified training data, which contain the weighted samples. New weak learners are added to the model sequentially to learn and identify tougher patterns. The data after every iteration is never the same and the possible misclassifications are pointed out for the algorithm to identify and learn from. The weights of the misclassifications are increased so that the next iteration can pick them up. The process is repeated for the specified number of iterations. The algorithm makes predictions based on the weak learner's majority vote coupled with their respective accuracies. AdaBoost is resistant to overfitting as the number of iterations increase and are most effective when it works on a binary classification problem. AdaBoost includes an extra condition where a model is required to have an error of less than 50% to maintain it, otherwise, the iterations are repeated until a better learner is generated.
```
Comment (srishilesh) [inline content/articles/boosting-algorithms-part-1/index.md:168]: This paragraph is too big. You could break it down into 2 - 3 paras.

File: content/articles/boosting-algorithms-part-1/index.md [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+
+![Adaboost](/engineering-education/boosting-algorithms-part-1/adaboost.png)
+
+*Source: [Towards Data Science](https://towardsdatascience.com/ensemble-methods-bagging-boosting-and-stacking-c9214a10a205)*
+
+AdaBoost for Regression works on the same principle, with the only difference being the predictions are made using the weighted average of the decision tree, with the weight being the accuracy of the learner against the training data. AdaBoost includes randomization in the construction of the models, so each time a piece of code is run, a slightly different model is generated. Stochastic learning algorithms produce different results every time the code is run, therefore, it is good practice to evaluate the performance of such algorithms by running the code a few times and taking the average of the results obtained.
```
Comment (srishilesh) [inline content/articles/boosting-algorithms-part-1/index.md:174]: Explain about `Stochastic learning algorithms`

File: content/articles/boosting-algorithms-part-1/index.md [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+
+*Source: [Machine Learning Mastery](https://machinelearningmastery.com/gentle-introduction-xgboost-applied-machine-learning/)*
+
+Breaking the process of boosting down from a mathematical standpoint, boosting is used to help find the minima of the 'n' features mapped in 'n' dimensional space, and most algorithms use gradient descent to find the minima. However, XGBoost uses a technique called the [Newton-Raphson method](https://en.wikipedia.org/wiki/Newton%27s_method), which provides a more direct route to the minima. Newton's Method uses the second derivative of the function which provides curvature information. This helps in the [faster minimization of functions](https://math.stackexchange.com/questions/1013195/why-is-newtons-method-faster-than-gradient-descent) when compared to [Gradient Descent](https://en.wikipedia.org/wiki/Gradient_descent), which uses the first derivative. In XGBoost, the decision trees that have nodes with weights that are generated with less evidence are shrunk heavily. This clever method of eliminating nodes that aren't significant is reflected in the speed of the algorithm and its memory consumption.
```
Comment (srishilesh) [inline content/articles/boosting-algorithms-part-1/index.md:204]: Break this paragraph down

File: content/articles/boosting-algorithms-part-1/index.md [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
...mes with an extra randomization parameter, which reduces the correlation between the trees. Less correlation between classifier trees translates to better performance of the ensemble of classifiers. XGBoost solves the problem of overfitting by correcting complex models with [regularization](https://www.section.io/engineering-education/regularization-to-prevent-overfitting/). Regularized Gradient Boosting is also available in XGBoost with both L1 and L2 regularization. For enhancing the processing performance, the algorithm uses multiple cores in the CPU. Multiple decision trees are generated in parallel by the algorithm by utilizing all the cores in your CPU. The system is designed as block-like structures, which enables the layout of the data to be reused in subsequent iterations instead of computing it all over again. This is used to find split points in the tree where the data points have equal weights, which makes it difficult to handle. The algorithm uses a distributed weighted quantile sketch algorithm to handle weighted data. The XGBoost library for Python is written in C++ and is available for C++, Python, R, Julia, Java, Hadoop and cloud-based platforms like AWS and Azure.
```
Comment (srishilesh) [inline content/articles/boosting-algorithms-part-1/index.md:210]: Break this paragraph down into 1 - 2 smaller paras

**Review by srishilesh — CHANGES_REQUESTED** [review]
@bloodprashure 

Great article. Very informative.

I've suggested a few changes. Please let me know when it's done.

Thank you.

**Comment by bloodprashure** [comment]
Hey @srishilesh, I've made the changes you requested. Do check them! Thanks!

**Review by srishilesh — APPROVED** [review]
@bloodprashure 

Thanks for making the changes.

Moving this article for the final review.

Thanks

**Comment by hectorkambow** [comment]
@bloodprashure 
[Boosting Algorithms in Python #2677.pdf](https://github.com/section-engineering-education/engineering-education/files/6727835/Boosting.Algorithms.in.Python.2677.pdf)

I'll be running a final review on your PR when ready.
Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept.
Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content.
Be sure to see our resources page to see more info on plagiarism and what is considered as such.

**Comment by bloodprashure** [comment]
Hello @hectorkambow. My apologies for the plagiarism report. On taking a look at the document, most similarities were from unrelated sources and the official documentation. However, I've corrected them now. Do take a look at it.
My apologies again, and I'll make sure to run my draft through a plagiarism checker before I submit it.
Thanks!

**Comment by bloodprashure** [comment]
@hectorkambow The draft's ready for its final review. Do let me know if any other changes should be made.
Thanks! 🚀

**Comment by hectorkambow** [comment]
great work gents 👍 
@bloodprashure

Please go through my final edits to see how future articles can be improved. 

cc @srishilesh Please do the same -some format edits that would help improve future reviews 🚀

**Comment by bloodprashure** [comment]
Thanks @hectorkambow! I'll make sure I format it right in future articles.
Thanks again! 🚀
